### ISSUE TITLE
Model can't train normally with TransformerBlock  

### ISSUE THREAD

**Issue by ytylalal** [ultralytics/ultralytics#934]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

When i add Transformer Block after C3 module( which could be normally using in V5), loss became nan(after increasing to 4~6), and training was abnormal.

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/500      15.1G        nan        nan        nan       1139        640: 100%|██████████| 94/94 [00:47<00:00,  2.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:02<00:00,  2.25it/s]
                   all        374      16177          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      2/500      15.1G        nan        nan        nan       1328        640: 100%|██████████| 94/94 [00:48<00:00,  1.93it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 6/6 [00:02<00:00,  2.23it/s]
                   all        374      16177          0          0          0          0

### Environment

windows 10  rtx3080laptop 

### Minimal Reproducible Example

class C3_VIT(nn.Module):
    # CSP Bottleneck with 3 convolutions
    def __init__(self, c1, c2, n=1, shortcut=True, g=1, e=0.5):  # ch_in, ch_out, number, shortcut, groups, expansion
        super(C3_VIT, self).__init__()
        c_ = int(c2 * e)  # hidden channels
        self.cv1 = Conv(c1, c_, 1, 1)
        self.cv2 = Conv(c1, c_, 1, 1)
        self.cv3 = Conv(2 * c_, c2, 1)  # optional act=FReLU(c2)
        self.m = nn.Sequential(*(Bottleneck(c_, c_, shortcut, g, e=1.0) for _ in range(n)))
        self.tr=TransformerBlock(c_,c_,4,1)

    def forward(self, x):
        return self.cv3(
            torch.cat(
                (self.tr(self.m(self.cv1(x))), self.cv2(x))
            , 1))

class TransformerBlock(nn.Module):
    # Vision Transformer https://arxiv.org/abs/2010.11929
    def __init__(self, c1, c2, num_heads, num_layers):
        super().__init__()
        self.conv = None
        if c1 != c2:
            self.conv = Conv(c1, c2)
        self.linear = nn.Linear(c2, c2)  # learnable position embedding
        self.tr = nn.Sequential(*[TransformerLayer(c2, num_heads) for _ in range(num_layers)])
        self.c2 = c2

    def forward(self, x):
        if self.conv is not None:
            x = self.conv(x)
        b, _, w, h = x.shape
        p = x.flatten(2)
        p = p.unsqueeze(0)
        p = p.transpose(0, 3)
        p = p.squeeze(3)
        e = self.linear(p)
        x = p + e

        x = self.tr(x)
        x = x.unsqueeze(3)
        x = x.transpose(0, 3)
        x = x.reshape(b, self.c2, w, h)
        return x

class TransformerLayer(nn.Module):
    # Transformer layer https://arxiv.org/abs/2010.11929 (LayerNorm layers removed for better performance)
    def __init__(self, c, num_heads):
        super().__init__()
        self.q = nn.Linear(c, c, bias=False)
        self.k = nn.Linear(c, c, bias=False)
        self.v = nn.Linear(c, c, bias=False)
        self.ma = nn.MultiheadAttention(embed_dim=c, num_heads=num_heads)
        self.fc1 = nn.Linear(c, c, bias=False)
        self.fc2 = nn.Linear(c, c, bias=False)

    def forward(self, x):
        x = self.ma(self.q(x), self.k(x), self.v(x))[0] + x
        x = self.fc2(self.fc1(x)) + x
        return x

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@ytylalal There could be various reasons why adding a Transformer Block after C3 module caused loss to become NaN during training. Here are some potential reasons and suggestions to fix the issue:

Check the hyperparameters: Adding a Transformer Block could change the number of parameters, so it's possible that the learning rate, batch size, or other hyperparameters need to be adjusted to prevent NaN loss. Try decreasing the learning rate and/or increasing the batch size.

Check the activation function: Make sure the activation function used in the Transformer Block is compatible with the rest of the network. ReLU is a common choice for most deep learning models.

Check the initialization method: NaN loss can occur if the weights of the Transformer Block are not initialized properly. Make sure the weights are initialized using a proper method, such as Xavier or He initialization.

Check for bugs in the code: Double-check the code for any bugs or typos that could be causing the issue. Print out the intermediate results and debug the code to see where the NaN loss is occurring.

Try removing the Transformer Block: If none of the above solutions work, it's possible that the Transformer Block is not compatible with the rest of the network. Try removing it and see if the training proceeds normally. If so, try using a different type of block that is more compatible with the rest of the network.

**Comment by rusvagzur** [comment]
Same happens to me when using the transformer block provided in yolov5

**Comment by AbdulManaf12** [comment]
I encountered NaN or zero values in the loss function when training the model with GPU. The issue was related to the GPU memory limitations, as the input data size was too large for the GPU memory.

Solution: I reduced the batch and input image sizes to reduce memory usage, and the issue was resolved. I recommend adjusting the batch and input sizes according to the available GPU memory to avoid out-of-memory errors and NaN values in the loss function.

**Comment by pderrenger** [comment]
@rusvagzur Thank you for sharing your experience. It appears that the issue with the transformer block causing NaN values during training is not unique to your specific implementation, as another user has experienced a similar issue using the transformer block provided in YOLOv5.

Considering that the block works with CPU but produces NaN values when using GPU, this might be related to GPU-specific optimizations or compatibility issues. To address this, I recommend investigating potential GPU-specific optimizations, double-checking GPU-specific dependencies, and ensuring that the GPU version of the framework and related libraries are up to date.

Furthermore, you may want to experiment with different GPU models and driver versions, as well as different PyTorch and CUDA versions to identify any specific compatibility issues.

Additionally, it would be valuable to review the CUDA toolkit and cuDNN versions being used, as they can have a significant impact on GPU performance and stability.

Please feel free to keep the conversation going, and I'll be happy to provide further assistance as needed.
